### PR TITLE
Migrate to AspNetCore

### PR DESCRIPTION
**Pull request by sprucely**
See https://github.com/ucdavis/Purchasing/projects/1 for status

### CODE REVIEW

File: Purchasing.Core/Services/IndexService.cs [hunk 1]
```diff
@@ -68,7 +71,17 @@ public ElasticSearchIndexService(IDbService dbService)
         {
             _dbService = dbService;
 
-            var pool = new SingleNodeConnectionPool(new Uri(ConfigurationManager.AppSettings["ElasticSearchUrl"]));
+            var match = Regex.Match(
```
[the remark this answers is not in the document]
Comment (jSylvestre) [reply Purchasing.Core/Services/IndexService.cs:74]: Maybe:
`Uri.TryCreate(System.Configuration.ConfigurationManager.AppSettings["ElasticSearchUrl"], UriKind.Absolute, out var uri);`

File: Purchasing.Jobs.CreateOrderIndexes/.upgrade-assistant [hunk 2]
```diff
@@ -0,0 +1 @@
+{"Build":"0.3.261602\u002Befb3cc30a24e699e6e4337f263a2b9478f04f618","CurrentProject":"Purchasing.Jobs.CreateOrderIndexes.csproj","EntryPoints":["Purchasing.Jobs.CreateOrderIndexes.csproj"],"Properties":{}}
```
Comment (jSylvestre) [inline Purchasing.Jobs.CreateOrderIndexes/.upgrade-assistant:1]: Remove this file?

File: Purchasing.Mvc/Attributes/AuthenticateListView.cs [hunk 3]
```diff
@@ -1,42 +1,52 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.Extensions.Configuration;
+using UCDArch.Core;
 
 namespace Purchasing.Mvc.Attributes
 {
     public class AuthenticateListView : ActionFilterAttribute
     {
+        public const string BLOCK_401_TO_302_REDIRECT = "Block401To302Redirect";
```
Comment (jSylvestre) [inline Purchasing.Mvc/Attributes/AuthenticateListView.cs:14]: We can't really test this unless we point sympa to this instance, but we don't really use this anyway....

File: Purchasing.Mvc/Attributes/AuthorizeOrderAccessAttribute.cs [hunk 4]
```diff
@@ -39,10 +41,8 @@ public override void OnAuthorization(AuthorizationContext filterContext)
 
             if (!_accessLevel.HasFlag(accessLevel))
             {
-                filterContext.Result = new HttpUnauthorizedResult(Resources.Authorization_PermissionDenied);
+                filterContext.Result = new UnauthorizedObjectResult(Resources.Authorization_PermissionDenied);
             }
-
-            base.OnAuthorization(filterContext);
```
Comment (jSylvestre) [inline Purchasing.Mvc/Attributes/AuthorizeOrderAccessAttribute.cs:45]: Was there anything behind the scenes being done by this deleted call?

Comment (jSylvestre) [inline Purchasing.Mvc/Attributes/AuthorizeOrderAccessAttribute.cs:45]: Was there anything behind the scenes being done by this deleted call?

File: Purchasing.Mvc/Attributes/IpFilter.cs [hunk 5]
```diff
@@ -1,24 +1,25 @@
...
         {
-            var clientIp = filterContext.RequestContext.HttpContext.Request.UserHostAddress;
+            var localIpFilter = SmartServiceLocator<IConfiguration>.GetService()["IpFilter"];
+            var clientIp = filterContext.HttpContext.Connection.RemoteIpAddress.MapToIPv4().ToString();
 
             // not authorized
             if (clientIp != localIpFilter)
             {
-                filterContext.Result = new RedirectToRouteResult(new RouteValueDictionary(new { controller = "error", action = "notauthorized" }));
+                filterContext.Result = new UnauthorizedObjectResult("You do not have access to this application");
```
Comment (jSylvestre) [inline Purchasing.Mvc/Attributes/IpFilter.cs:22]: This isn't really the same, but it doesn't look like we use this either, so no worries.

File: Purchasing.Mvc/AutomapperInstaller.cs [hunk 6]
```diff
@@ -1,22 +1,42 @@
...
+    public class AutoMapperInstaller : IWindsorInstaller
     {
-        public static void Configure()
+        public void Install(IWindsorContainer container, IConfigurationStore store)
         {
-            Mapper.Initialize(cfg => cfg.AddProfile<ViewModelProfile>());
+            // Register all mapper profiles
+            container.Register(
+                Classes.FromAssemblyInThisApplication(GetType().Assembly)
+                .BasedOn<Profile>().WithServiceBase());
```
Comment (jSylvestre) [inline Purchasing.Mvc/AutomapperInstaller.cs:18]: Profile? Should this be ViewModelProfile ?

Comment (sprucely) [reply Purchasing.Mvc/AutomapperInstaller.cs:18]: Yes. Good catch.

Comment (sprucely) [reply Purchasing.Mvc/AutomapperInstaller.cs:18]: NVM, this is how the Automapper docs say to do it.

File: Purchasing.Mvc/Controllers/AdminController.cs [hunk 7]
```diff
@@ -573,7 +589,7 @@ public ActionResult ValidateChildWorkgroups()
         public virtual bool NeedToCheckWorkgroupPermissions(string key)
```
Comment (jSylvestre) [inline Purchasing.Mvc/Controllers/AdminController.cs:589]: We don't call this from anywhere anymore. It is very database intensive.
Possibly we should remove it.

File: Purchasing.Mvc/Controllers/AjaxController.cs [hunk 8]
```diff
@@ -1,21 +1,19 @@
...
 using UCDArch.Web.ActionResults;
 using IdAndName = Purchasing.Core.Services.IdAndName;
+using Microsoft.AspNetCore.Mvc;
 
 namespace Purchasing.Mvc.Controllers
 {
     /// <summary>
     /// Controller for AJAX service calls
     /// </summary>
-    [SessionState(SessionStateBehavior.Disabled)] //Disabled so requests can run in parallel
```
Comment (jSylvestre) [inline Purchasing.Mvc/Controllers/AjaxController.cs:18]: Searching for a building works, but I don't know if these changes will break something else.

File: Purchasing.Mvc/Controllers/ConditionalApprovalController.cs [hunk 9]
```diff
@@ -85,7 +84,7 @@ public ActionResult ByWorkgroup(int id)
         /// <returns></returns>
         public ActionResult Delete(int id)
         {
-            ActionResult redirectToAction;
+            Microsoft.AspNetCore.Mvc.ActionResult redirectToAction;
```
Comment (jSylvestre) [inline Purchasing.Mvc/Controllers/ConditionalApprovalController.cs:87]: This change can be removed. It was added in the usings

File: Purchasing.Mvc/Controllers/OrderController.cs [hunk 10]
```diff
@@ -1201,10 +1203,9 @@ public ActionResult AddVendor(int workgroupId, WorkgroupVendor vendor)
 
             Check.Require(_securityService.HasWorkgroupAccess(workgroup));
 
-            var modelState = new ModelStateDictionary();
```
Comment (jSylvestre) [inline Purchasing.Mvc/Controllers/OrderController.cs:1204]: @sprucely I think this changes the functionality and should probably be reverted.

File: Purchasing.Mvc/Models/WorkgroupModifyModel.cs [hunk 11]
```diff
@@ -15,22 +15,22 @@ namespace Purchasing.Mvc.Models
     public class WorkgroupModifyModel
     {
         public Workgroup Workgroup { get; set; }
-        public List<ListItem> Organizations { get; set; } 
+        public List<SelectListItem> Organizations { get; set; }
```
Comment (jSylvestre) [inline Purchasing.Mvc/Models/WorkgroupModifyModel.cs:18]: @sprucely Could this change be why it is displaying differently?

File: Purchasing.Jobs.NotificationsCommon/ProcessNotifications.cs [hunk 12]
```diff
@@ -150,39 +151,35 @@ private static void BatchEmail(IDbConnection connection, string email, List<dyna
 
                 message.Append(string.Format("<p><em>{0} </em><em><a href=\"{1}\">{2}</a>&nbsp;</em></p>", "You can change your email preferences at any time by", "http://prepurchasing.ucdavis.edu/User/Profile", "updating your profile on the PrePurchasing site"));
 
-                var emailTransmission = new Transmission
+                using var client = new SmtpClient("smtp.sparkpostmail.com", 587)
```
Comment (jSylvestre) [inline Purchasing.Jobs.NotificationsCommon/ProcessNotifications.cs:154]: use config settings

File: Purchasing.Jobs.NotificationsCommon/ProcessNotifications.cs [hunk 13]
```diff
@@ -150,39 +151,35 @@ private static void BatchEmail(IDbConnection connection, string email, List<dyna
...
+                    Body = message.ToString(),
+                    IsBodyHtml = true,
+                };
+                mailMessage.To.Add(new MailAddress(email));
 
-                var client = new Client(SparkPostApiKey);
                 try
                 {
-                    client.Transmissions.Send(emailTransmission).Wait();
+                    client.Send(mailMessage);
```
Comment (jSylvestre) [inline Purchasing.Jobs.NotificationsCommon/ProcessNotifications.cs:174]: awaitable?

File: Purchasing.Jobs.NotificationsCommon/Purchasing.Jobs.NotificationsCommon.csproj [hunk 14]
```diff
@@ -1,110 +1,29 @@
...
-  </Target>
-  <Target Name="AfterBuild">
-  </Target>
-  -->
+	<ItemGroup>
+		<ProjectReference Include="..\Purchasing.Core\Purchasing.Core.csproj"/>
+		<ProjectReference Include="..\Purchasing.Jobs.Common\Purchasing.Jobs.Common.csproj"/>
+	</ItemGroup>
+	<ItemGroup>
+		<PackageReference Include="Sendgrid" Version="9.27.0"/>
```
Comment (jSylvestre) [inline Purchasing.Jobs.NotificationsCommon/Purchasing.Jobs.NotificationsCommon.csproj:23]: do we need sendgrid?

**Review by jSylvestre — CHANGES_REQUESTED** [review]

File: Purchasing.Mvc/Controllers/SearchController.cs [hunk 15]
```diff
@@ -57,7 +54,7 @@ public ActionResult Results(string q)
             {
                 if (Repository.OfType<Order>().Queryable.Any(a => a.RequestNumber == q.Trim().ToUpper()))
                 {
-                    return this.RedirectToAction<OrderController>(a => a.Lookup(q.Trim().ToUpper()));
+                    return this.RedirectToAction(nameof(OrderController.Lookup), typeof(OrderController).ControllerName(), new { requestNumber = q.Trim().ToUpper() });
```
[the remark this answers is not in the document]
Comment (jSylvestre) [reply Purchasing.Mvc/Controllers/SearchController.cs:57]: nm, I see now.
This is fixed with pr #489

File: Purchasing.Mvc/Views/Shared/_JsBundle.cshtml [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+      <script src="~/Scripts/public/order/SausageCustom.js"></script>
+    </environment>
+    break;
+
+  case "~/bundles/orderreview":
+    <environment names="Staging,Production">
+      <script src="~/bundles/orderreview.min.js"></script>
+    </environment>
+    <environment names="Development">
+      <script src="~/Scripts/public/single/fileuploader.js"></script>,
```
Comment (jSylvestre) [inline Purchasing.Mvc/Views/Shared/_JsBundle.cshtml:81]: @sprucely Trailing commas. I'll remove them.

File: Purchasing.Mvc/Views/Order/_OrderDetails.cshtml [hunk 17]
```diff
@@ -56,7 +56,7 @@
                     <h2>Select an approver &amp; account manager</h2>
                     <div class="editor-label">Select An Approver:</div>
                     <div class="editor-field">
-                        @this.Select("approvers").FirstOption("--Select An Approver (Optional)--").Options(Model.Approvers, x => x.Id, x => x.FullNameAndId)
+                        @Html.Select("approvers", o => o.FirstOption("--Select An Approver (Optional)--").Options(Model.Approvers, x => x.Id, x => x.FullNameAndId))
```
Comment (jSylvestre) [inline Purchasing.Mvc/Views/Order/_OrderDetails.cshtml:59]: If we want this to order correctly, do it like the account managers select list below.
FirstOption isn't making it the first option here

File: Purchasing.Mvc/Program.cs [hunk 18]
```diff
@@ -0,0 +1,65 @@
+﻿// This application entry point is based on ASP.NET Core new project templates and is included
+// as a starting point for app host configuration.
+// This file may need updated according to the specific scenario of the application being upgraded.
+// For more information on ASP.NET Core hosting, see https://docs.microsoft.com/aspnet/core/fundamentals/host/web-host
+
```
Comment (srkirkland) [inline Purchasing.Mvc/Program.cs:5]: dump

File: Purchasing.Mvc/Startup.cs [hunk 19]
```diff
@@ -0,0 +1,175 @@
+﻿// This Startup file is based on ASP.NET Core new project templates and is included
+// as a starting point for DI registration and HTTP request processing pipeline configuration.
+// This file will need updated according to the specific scenario of the application being upgraded.
+// For more information on ASP.NET Core startup files, see https://docs.microsoft.com/aspnet/core/fundamentals/startup
+
```
Comment (srkirkland) [inline Purchasing.Mvc/Startup.cs:5]: dump this too

File: Purchasing.Mvc/Startup.cs [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+            app.UseMiddleware<CorrelationIdMiddleware>();
+
+            if (env.IsDevelopment())
+            {
+                app.UseDeveloperExceptionPage();
+            }
+            else
+            {
+                app.UseExceptionHandler("/Error/Index");
+                app.UseStatusCodePages(context =>
```
Comment (srkirkland) [inline Purchasing.Mvc/Startup.cs:140]: this is pretty cool, haven't seen it before

File: Purchasing.Mvc/bundleConfig.json [hunk 21]
```diff
@@ -0,0 +1,132 @@
+[
```
Comment (srkirkland) [inline Purchasing.Mvc/bundleConfig.json:1]: general question: does the bundleConfig still work/is still being used in net6 apps?

File: .config/dotnet-tools.json [hunk 22]
```diff
@@ -0,0 +1,12 @@
+{
+  "version": 1,
+  "isRoot": true,
+  "tools": {
+    "bundlerminifier.core.tool": {
```
Comment (srkirkland) [inline .config/dotnet-tools.json:5]: ignore my earlier comment about the bundleConfig file -- i see this is how it's all tied together

**Review by srkirkland — APPROVED** [review]
Great job, I made a bunch of comments but they are all minor.  I skipped reviewing the UCDArch & Testing pieces since it was hard to tell what changed there.  Overall looks really good to me, glad other than the select boxes most changes didn't affect core controllers, db queries or views, so that gives me some confidence.

File: Purchasing.Mvc/Views/Order/_OrderDetails.cshtml (hunk 17 above)
Comment (sprucely) [reply Purchasing.Mvc/Views/Order/_OrderDetails.cshtml:59]: Nah, just a quick fix in `HtmlHelperSelectExtensions.cs`

**Pull request merged**